lib/Lexer/PHP.hs

```diff
@@ -0,0 +1,121 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# LANGUAGE QuasiQuotes, OverloadedStrings #-}
+module Lexer.PHP() where
+
+import qualified Data.List.NonEmpty as NE
+import Regex.Parse
+import Data.Maybe
+import Data.Text (Text)
+import qualified Data.Text as T
+import Lexer.Types
+import Lang
+import Utils
+
+instance LexerWriter PHP where
+  writeLexer _ accSt tokNames stList =
+    [ ("Lexer.php", [interp|
+<?php
+
+namespace Alpaca;
+
+class Lexer
+{
+    const TOKEN_TYPE_EOF = 0;
+    #{indent 1 tokDefns}
+
+    private $input;
+    private $curChIx;
+    private $debug;
+
+    public function __construct(string $input, int $curChIx, bool $debug)
+    {
+        $this->input = $input;
+        $this->curChIx = $curChIx;
+        $this->debug = $debug;
+    }
+
+    public static function tokenToString(int $token): string
+    {
+        switch($token) {
+            case 0: return '%eof';
+            #{indent 3 tokToStr}
+            default: throw new \\InvalidArgumentException("Unknown token $token");
+        }
+    }
+
+    public function getNextToken(): array
+    {
+        $lastAccChIx = $this->curChIx;
+        $startChIx = $this->curChIx;
+        $accSt = -1;
+        $curSt = 0;
+        while ($curSt >= 0) {
+        if (in_array($curSt, [#{T.intercalate "," $ map (tshow . fst) accSt}])) {
+                $lastAccChIx = $this->curChIx;
+                $accSt = $curSt;
+            }
+
+            if ($this->curChIx >= strlen($this->input))
+                break;
+
+            $curCh = $this->input[$this->curChIx];
+            $this->curChIx+=1;
+            switch($curSt) {
+                #{indent 4 transTable}
+            }
+            break;
+        }
+
+        $lastReadChIx = $this->curChIx;
+        $this->curChIx = $lastAccChIx;
+        $text = substr($this->input, $startChIx, $lastAccChIx);
```

Comment:
In JS, `substr` takes `string`, `startIndex`, `endIndex`. In PHP, it takes `string`, `startIndex`, `length`. This leads to a bug, since you're passing index as length.